jso/apis/src/main/java/org/teavm/jso/core/JSPromise.java

```diff
@@ -0,0 +1,71 @@
+/*
+ *  Copyright 2023 Bernd Busse.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.teavm.jso.core;
+
+import org.teavm.interop.NoSideEffects;
+import org.teavm.jso.JSBody;
+import org.teavm.jso.JSByRef;
+import org.teavm.jso.JSFunctor;
+import org.teavm.jso.JSMethod;
+import org.teavm.jso.JSObject;
+
+public abstract class JSPromise implements JSObject {
+    private JSPromise() {
+    }
+
+    @JSFunctor
+    public interface Executor extends JSObject {
+        void onExecute(JSFunction resolveFunc, JSFunction rejectFunc);
```

Comment:
I'd use a more specific interface, like `JSConsumer<T>` for this purpose. I know, it's a bit inconvenient that `java.util.function` interfaces can't be used in this context. I'll consider to design support of these functional interfaces in JSO. For now you can declare something like `JSConsumer<T>` yourself and declare `resolveFunc` as `JSConsumer<T>`, and `rejectFunc` as `JSConsumer<Object>`

Reply:
Just to clarify whether I understand this correctly:
You'd prefer a JSO specific "copy" or extension of the `java.utils.function.Consumer` interface (for now), and use that as the type for the JavaScript supplied `resolveFunc` and `rejectFunc`? This is to be more specific about their signature (1 parameter, no return)?

Can't this `Executor` interface be a (custom) `BiConsumer` then, or does this still need more glue code in the code generation path?

What about the `then`, `catch` and `finally` methods? Is the `JSFunction` type correct for their arguments to pass Java functions to the JavaScript context, or would these also benefit from a more specific interface?

Reply:
You can replace `Executor` with a custom `BiConsumer`, but this results in a hard-to-understand signature like: `BiConsumer<Consumer<T>, Consumer<Object>>`, which is not quite readable and default parameter names won't be quite descriptive.

For `then`, `catch` and `finally` it's also necessary to introduce dedicated interfaces. `JSFunction` does not fit for these methods at all, since it's not a functional interface (i.e. does not provide exactly one method). It was designed for opposite case: *importing* JavaScript functions into Java (at least, the 'general' one, when developer is lazy enough to provide correct declaration with proper types and descriptive name). Please also note that `then` can return not only value itself, but Promise that resolves into value. In Java it's not possible to express due to lack of union types, so I suggest following: add another version of `then` method (e.g. `flatThen` or `thenPromise`) that expected to return promise. This won't solve *all* possible cases for JS promises like different branches of `if` inside lambda returning both value and promise, but this trade-off looks at least good enough.